### PR TITLE
Implement access to `device` and `dtype`

### PR DESCRIPTION
**Pull request by SCiarella**
Closes issue #23

This PR exposes `device` and `dtype` such that they can be controlled and set externally by the user. This is particularly relevant for running `diffwofost` on GPU by setting `device="cuda"`. There are example of this in the notebooks and in the documentation.

The tests have been update by adding a new `@pytest.fixture` that checks if a GPU is available, and if so, it re-runs the most important tests on the GPU a second time. Notice that GitHub actions do not provide GPU access (for free), so I have been using my local hardware to test the code.

-----

Notice that without finalizing the Engine, the GPU device could be slower than the CPU, due to temporary fixes like the use of `deepcopy` in the `TestEngine`.

### CODE REVIEW

File: docs/notebooks/optimization.ipynb
Comment (fnattino) [inline docs/notebooks/optimization.ipynb]: The notebook fails because of the changes introduced in #59 (the engine now allows in-memory configuration, so the .conf files have been removed). The fix will be included as part of #69, so we can drop the changes to this notebook.

File: pyproject.toml [hunk 1]
```diff
@@ -61,6 +61,9 @@ Issues = "https://github.com/WUR-AI/diffwofost/issues"
 
 [tool.pytest.ini_options]
 testpaths = ["tests"]
+filterwarnings = [
```
Comment (fnattino) [inline pyproject.toml]: Oh, so clean this can be configured here!

File: tests/physical_models/crop/test_leaf_dynamics.py [hunk 2]
```diff
@@ -130,7 +139,10 @@ def test_leaf_dynamics_with_one_parameter_vector(self, param):
             weather_data_provider,
             agro_management_inputs,
             external_states,
-        ) = prepare_engine_input(test_data, crop_model_params, meteo_range_checks=False)
+        ) = prepare_engine_input(
+            test_data, crop_model_params, meteo_range_checks=False, device=device
+        )
+        # config_path = str(phy_data_folder / "WOFOST_Leaf_Dynamics.conf")
```
Comment (fnattino) [inline tests/physical_models/crop/test_leaf_dynamics.py]:
```suggestion
```

File: tests/physical_models/crop/test_phenology.py [hunk 3]
```diff
@@ -74,23 +78,28 @@ def __init__(
...
         self.weather_data_provider = weather_data_provider
         self.agro_management_inputs = agro_management_inputs
         self.config = config
+        self.device = device
 
     def forward(self, params_dict):
         # pass new value of parameters to the model
         for name, value in params_dict.items():
+            if isinstance(value, torch.Tensor) and value.device.type != self.device:
+                value = value.to(self.device)
```
Comment (fnattino) [inline tests/physical_models/crop/test_phenology.py]: Same as for `test_leaf_dynamics`:
```suggestion
```

File: tests/physical_models/crop/test_root_dynamics.py [hunk 4]
```diff
@@ -420,23 +438,31 @@ class TestDiffRootDynamicsGradients:
...
         value, dtype = self.param_configs[config_type][param_name]
-        param = torch.nn.Parameter(torch.tensor(value, dtype=dtype))
+        param = torch.nn.Parameter(torch.tensor(value, dtype=dtype, device=device))
         output = model({param_name: param})
         loss = output[output_name].sum()
 
-        assert loss.grad_fn is None
+        # If there is no graph at all, then there is no gradient by definition.
+        if not loss.requires_grad:
+            return
```
Comment (fnattino) [inline tests/physical_models/crop/test_root_dynamics.py]: Uhm, this conditional essentially passes without any assertion. In order to make this test more effective, should one split the parameters that we know give an empty graph from the parameters that give a zero gradient, and assert that the right situation is encountered?

File: src/diffwofost/physical_models/crop/phenology.py [hunk 5]
```diff
@@ -417,9 +533,46 @@ def initialize(self, day, kiosk, parvalues):
...
         self.params_shape = _get_params_shape(self.params)
 
+        # Small epsilon tensor reused in multiple safe divisions.
+        self._epsilon = torch.tensor(1e-8, dtype=self.dtype, device=self.device)
+
+        # Ensure the Vernalisation submodule uses the same dtype/device as this phenology instance
+        Vernalisation.device = self.device
+        Vernalisation.dtype = self.dtype
+
+        # Helpler function to cast and broadcast all parameters to params_shape.
```
Comment (fnattino) [inline src/diffwofost/physical_models/crop/phenology.py]: Does this function need to be a closure or can it be written as a regular method?
 
Spotted a typo:
```suggestion
        # Helper function to cast and broadcast all parameters to params_shape.
```

File: tests/physical_models/crop/test_leaf_dynamics.py [hunk 6]
```diff
@@ -46,17 +44,21 @@ def __init__(
...
         self.agro_management_inputs = agro_management_inputs
         self.config = config
         self.external_states = external_states
+        self.device = device
 
     def forward(self, params_dict):
         # pass new value of parameters to the model
         for name, value in params_dict.items():
+            if isinstance(value, torch.Tensor) and value.device.type != self.device:
+                value = value.to(self.device)
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply tests/physical_models/crop/test_leaf_dynamics.py]: not needed with centralized control

**Comment by SCiarella** [comment]
Thanks @fnattino for the careful review!

I have implemented most of the suggestions that you proposed. In particular, I introduced a new class in `config.py` called `ComputeConfig` that is used to set globally `dtype` and `device`. Using this global control, I was able to remove most of the redundant type checks that you also pointed out. @SarahAlidoost what do you think of this solution?

There is a point that remains open regarding the `kiosk`. Currently, I have implemented some type/device conversion for the variables that come from the kiosk. Later, we should ensure that the engine enforces global types and devices, such that the kiosk does not need to be converted every time.
